hpx/async.hpp

```diff
@@ -101,12 +101,12 @@ namespace hpx { namespace detail
         template <typename F, typename ...Ts>
         BOOST_FORCEINLINE static
         typename boost::enable_if_c<
-            traits::detail::is_deferred_callable<F(Ts...)>::value,
-            hpx::future<typename util::deferred_call_result_of<F(Ts...)>::type>
+            traits::detail::is_deferred_callable<F(Ts&&...)>::value,
+            hpx::future<typename util::detail::deferred_result_of<F(Ts&&...)>::type>
         >::type
         call(BOOST_SCOPED_ENUM(launch) launch_policy, F&& f, Ts&&... ts)
         {
-            typedef typename util::deferred_call_result_of<
+            typedef typename util::detail::deferred_result_of<
                 F(Ts...)
```

Comment:
On 11/23/2015 01:00 PM, Agustín Bergé wrote:
> In hpx/async.hpp
> <https://github.com/STEllAR-GROUP/hpx/pull/1875#discussion_r45595802>:
> 
>>          >::type
>>          call(BOOST_SCOPED_ENUM(launch) launch_policy, F&& f, Ts&&... ts)
>>          {
>> -            typedef typename util::deferred_call_result_of<
>> +            typedef typename util::detail::deferred_result_of<
>>                  F(Ts...)
> 
> Parameter types in function types are decayed. When doing perfect
> forwarding, forgetting the |&&| means that for rvalues qualifiers will
> be lost, and things might spontaneously turn into pointers. They should
> all be |Ts&&...|, I missed this one.

Good to know, thanks.